algorithms/kernel/svm/oneapi/cl_kernels/svm_train_block_smo_oneapi.cl

```diff
@@ -0,0 +1,234 @@
+/* file: svm_train_block_smo_oneapi.cl */
+/*******************************************************************************
+* Copyright 2020 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+/*
+//++
+//  Implementation of SMO algorithm for wokset block.
+//--
+*/
+
+#ifndef __SVM_TRAIN_BLOCK_SMO_ONEAPI_CL__
+#define __SVM_TRAIN_BLOCK_SMO_ONEAPI_CL__
+
+#include <string.h>
+
+#define DECLARE_SOURCE_DAAL(name, src) static const char *(name) = #src;
+
+DECLARE_SOURCE_DAAL(
+    clKernelBlockSMO,
+
+    inline bool isUpper(const algorithmFPType alpha, const algorithmFPType y, const algorithmFPType C) {
+        return (y > 0 && alpha < C) || (y < 0 && alpha > 0);
+    }
+
+    inline bool isLower(const algorithmFPType alpha, const algorithmFPType y, const algorithmFPType C) {
+        return (y > 0 && alpha > 0) || (y < 0 && alpha < C);
+    }
+
+    void reduceMax(const __local algorithmFPType * values, __local int * indices) {
+        const int local_id = get_local_id(0);
```

Comment:
You probably can use `sub_group_reduce_max()` from https://www.khronos.org/registry/OpenCL/extensions/intel/cl_intel_subgroups.html

Reply:
Done. Thanks! It It was an interesting experience.